// Copyright 2021 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package promise_fuzzer;

message Seq {
    Promise first = 1;
    repeated PromiseFactory promise_factories = 2;
}

message Join {
    repeated Promise promises = 1;
}

message Race {
    repeated Promise promises = 1;
}

message PromiseFactory {
    oneof promise_factory_type {
        Promise promise = 1;
    }
}

message Never {}

message Promise {
    oneof promise_type {
        Seq seq = 1;
        Join join = 2;
        Race race = 3;
        Never never = 4;
        int32 sleep_first_n = 5;
    }
}

message Cancel {}

message Wakeup {}

message Action {
    oneof action_type {
        Wakeup wakeup = 1;
        Cancel cancel = 2;
        int32 run_todo = 3;
    }
}

message Msg {
    Promise promise = 1;
    repeated Action actions = 2;
}
